### PR TITLE
[2.0] Remove mutability of geometries 

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
I branched off from master, so the shapely-2.0 branch first needs to be reset to master state for this to work (the actual commit is only the last one)

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@sgillies could you rebase the `shapely-2.0` branch on top of master? (see also #962) Then the diff of this PR should be more correct.

**Comment by sgillies** [comment]
@jorisvandenbossche done. You still need to rebase your own branch, right?

**Comment by jorisvandenbossche** [comment]
Thnks! Switching branches back and again also seemed to trigger GitHub

**Comment by jorisvandenbossche** [comment]
@sgillies @mwtoews this is ready for review/merge

File: shapely/geometry/base.py [hunk 1]
```diff
@@ -338,11 +338,7 @@ def _get_coords(self):
             return []
         return CoordinateSequence(self)
 
-    def _set_coords(self, ob):
-        raise NotImplementedError(
-            "set_coords must be provided by derived classes")
```
Comment (sgillies) [inline shapely/geometry/base.py:343]: Bye bye fragile mutable geometries :hocho:

File: tests/test_linestring.py [hunk 2]
```diff
@@ -132,16 +132,6 @@ def test_linestring(self):
                          {'type': 'LineString',
                           'coordinates': ((1.0, 2.0), (3.0, 4.0))})
 
-    @shapely20_deprecated
-    def test_linestring_mutate(self):
```
Comment (mwtoews) [inline tests/test_linestring.py:136]: Perhaps rename this `test_linestring_immutable` and check that `.coord` raise TypeError.

File: tests/test_point.py [hunk 3]
```diff
@@ -113,19 +113,6 @@ def test_point(self):
         self.assertEqual(p.__geo_interface__,
                          {'type': 'Point', 'coordinates': (3.0, 4.0)})
 
-    @shapely20_deprecated
-    def test_point_mutate(self):
```
Comment (mwtoews) [inline tests/test_point.py:117]: Similar recommendation as above: rename this `test_point_immutable` and check that `.coord` raise TypeError.

**Review by mwtoews — COMMENTED** [review]
Theses changes are welcome! I have suggestions to keep two tests to check that point and linestring geometries now have immutable `.coords` properties. It might be helpful to the reader to see a comment to say that these were mutable in previous versions. All other tests should be dropped, however.

**Comment by jorisvandenbossche** [comment]
> keep (and modify) two tests to check that point and linestring geometries now have immutable .coords properties. It might be helpful to the reader to see a comment to say that these were mutable in previous versions. 

Good point, will update for that. 

Thanks for the review!

**Comment by jorisvandenbossche** [comment]
Added back tests for point/linestring/linearring to ensure immutability.

We now have those errors:

```python
In [1]: from shapely.geometry import LineString

In [3]: l = LineString([(0,0), (1,1)])


In [6]: l.coords[0] = (5, 5)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-6-41687dadde5d> in <module>
----> 1 l.coords[0] = (5, 5)

TypeError: 'CoordinateSequence' object does not support item assignment

In [7]: l.coords = [(0,0), (1,1)]
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-7-dc6ae96a6b61> in <module>
----> 1 l.coords = [(0,0), (1,1)]

AttributeError: can't set attribute
```

which are the default errors produced by Python for a property without a setter (the first one about CoordinateSequence is already existing on master, just added to the test for completeness). 

Are we OK with that test? I could also add back a setter to the `coords` property with a more custom error if we want that.

**Review by mwtoews — APPROVED** [review]
:tada:

**Comment by sgillies** [comment]
@jorisvandenbossche I'm fine with that. Since we're going to have a warning period, a custom error won't be necessary. If these objects had custom setters, custom exceptions would help users differentiate between Shapely usage bugs and Shapely bugs/Python usage bugs. In 2.0, trying to modify a coordinate sequence is a Python usage bug, and a type or attribute error fits.

**Comment by jorisvandenbossche** [comment]
OK, this should be good to go then!

**Comment by sgillies** [comment]
@mwtoews @jorisvandenbossche merge when you are ready. I'm happy to let the two of you be stewards of the shapely-2.0 branch while I focus on the others.

**Comment by jorisvandenbossche** [comment]
I would be happy with managing the shapely-2.0 branch (as noted in https://github.com/Toblerity/Shapely/issues/962, I would also prefer to regularly rebase it on top of master initially), but then you will need to give me commit rights ;) (I can ensure to use them wisely!)

**Comment by sgillies** [comment]
@jorisvandenbossche I'd lost track of that requirement and will make it so.

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**